R/coalesce_linter.R

```diff
@@ -0,0 +1,87 @@
+#' Encourage usage of the null coalescing operator `%||%`
+#'
+#' The `x %||% y` is equivalent to
+#'   `if (is.null(x)) y else x`, but more expressive.
+#'   It is exported by R since 4.4.0, and equivalents
+#'   have been available in other tidyverse packages
+#'   for much longer, e.g. 2008 for ggplot2.
+#'
+#' @examples
+#' # will produce lints
+#' lint(
+#'   text = "if (is.null(x)) y else x",
+#'   linters = coalesce_linter()
+#' )
+#'
+#' lint(
+#'   text = "if (!is.null(x)) x else y",
+#'   linters = coalesce_linter()
+#' )
+#'
+#' lint(
+#'   text = "if (is.null(x[1])) x[2] else x[1]",
+#'   linters = coalesce_linter()
+#' )
+#'
+#' # okay
+#' lint(
+#'   text = "x %||% y",
+#'   linters = coalesce_linter()
+#' )
+#'
+#' lint(
+#'   text = "x %||% y",
+#'   linters = coalesce_linter()
+#' )
+#'
+#' lint(
+#'   text = "x[1] %||% x[2]",
+#'   linters = coalesce_linter()
+#' )
+#'
+#'
+#' @evalRd rd_tags("coalesce_linter")
+#' @seealso [linters] for a complete list of linters available in lintr.
+#' @export
+coalesce_linter <- function() {
+  braced_expr_cond <- "expr[1][OP-LEFT-BRACE and count(*) = 3]/expr"
+  xpath <- glue("
```

Comment:
Anchor on `IF` for better performance?

Reply:
it means changing from the cached function call approach to a full `//IF` search. I guess the current approach is better, or at least not a notable difference (`is.null()` is not a ubiquitous function like `c()`).

If you think it'll be a lot more readable I can try and make the swtich.